msgraph-developer-proxy-plugins/Behavior/RetryAfterPlugin.cs

```diff
@@ -0,0 +1,102 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using Microsoft.Extensions.Configuration;
+using Microsoft.Graph.DeveloperProxy.Abstractions;
+using System.Net;
+using System.Text.Json;
+using System.Text.RegularExpressions;
+using Titanium.Web.Proxy.Http;
+using Titanium.Web.Proxy.Models;
+
+namespace Microsoft.Graph.DeveloperProxy.Plugins.Behavior;
+
+public class RetryAfterPlugin : BaseProxyPlugin
+{
+  public override string Name => nameof(RetryAfterPlugin);
+
+  public override void Register(IPluginEvents pluginEvents,
+                       IProxyContext context,
+                       ISet<UrlToWatch> urlsToWatch,
+                       IConfigurationSection? configSection = null)
+  {
+    base.Register(pluginEvents, context, urlsToWatch, configSection);
+
+    pluginEvents.BeforeRequest += OnRequest;
+  }
+
+  private async Task OnRequest(object? sender, ProxyRequestArgs e)
+  {
+    if (e.ResponseState.HasBeenSet ||
+        _urlsToWatch is null ||
+        !e.ShouldExecute(_urlsToWatch))
+    {
+      return;
+    }
+
+    ThrottleIfNecessary(e);
+  }
+
+  private void ThrottleIfNecessary(ProxyRequestArgs e)
+  {
+    var request = e.Session.HttpClient.Request;
+    var expiredThrottlers = e.ThrottledRequests.Where(t => t.ResetTime < DateTime.Now);
+    foreach (var throttler in expiredThrottlers)
+    {
+      e.ThrottledRequests.Remove(throttler);
+    }
+
+    if (!e.ThrottledRequests.Any())
+    {
+      return;
+    }
+
+    foreach (var throttler in e.ThrottledRequests)
+    {
+      var throttleInfo = throttler.ShouldThrottle(request, throttler.ThrottlingKey);
+      if (throttleInfo.ThrottleForSeconds > 0)
+      {
+        _logger?.LogRequest(new[] { $"Calling {request.Url} before waiting for the Retry-After period.", "Request will be throttled", $"Throttling on {throttler.ThrottlingKey}" }, MessageType.Failed, new LoggingContext(e.Session));
+
+        throttler.ResetTime = throttler.ResetTime.AddSeconds(throttleInfo.ThrottleForSeconds);
```

Comment:
ThrottleForSeconds needs to be set to DateTime.Now - ResetTime. Right now, we incorrectly manage expectations where we tell you to back off for 5 seconds while we're expecting you to wait reset time + 5 which leads to force fails all the time.